articles/external_qos_file.md

```diff
@@ -0,0 +1,658 @@
+        </action_client>
+    </node>
+</qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        action_client:
+            name: my_action
+            feedback:
+                qos:
+                    history_depth: 100
+```
+
+It would be worth revisiting if having 5 different QoS profiles for defining actions make sense before extending the file format in such a way.
+
+### Support in launch files
+
+`launch_ros.actions.Node` and `launch_ros.actions.LoadComposableNode` actions could support a qos file argument in their constructors.
+
+### rmw implementation specific payload
+
+There are currently QoS settings that are supported by DDS and not by ROS.
+An user might want to make use of them, and they won't find an easy way to do that without exposing the QoS setting on ROS.
+
+There are some cases where an user might want to use a vendor specific QoS extension, for example rti allows defining [flow controllers](https://community.rti.com/static/documentation/connext-dds/5.2.0/doc/manuals/connext_dds/html_files/RTI_ConnextDDS_CoreLibraries_UsersManual/Content/UsersManual/CreatingandConfigCustomFlowC.htm) in publishers.
+
+Though a DDS vendor specific QoS profile file can be used to set up these QoS settings, you currently can only use the default qos profile.
+There's no way to match ROS entities to the DDS QoS profile names.
+
+This problem can be solved in two ways:
+- Having conventions to match DDS QoS profiles names to ROS entities.
+  This has been proposed in https://github.com/ros2/rmw_fastrtps/pull/335, and could work well for most use cases.
+- Adding a rmw implementation specific "payload" that it's then used by the rmw implementation.
+  In DDS based implementations, this payload can be just the profile name.
+
+As an example of the second alternative:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        <publisher topic_name="asd">
+            <qos>
+                <rmw_payload>
+                    <profile_name>my_qos_profile_name</profile_name>
+                </rmw_payload>
+            </qos>
+        </publisher>
+    </node>
+</qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        publisher:
+            topic_name: asd
+            rmw_payload:
+                profile_name: my_qos_profile_name
+```
+
+For this to work, `rmw_qos_profile_t` will need to be extended with a `rmw_qos_profile_payload_t` member defined like:
+
+```c
+typedef struct rmw_qos_profile_payload_item_t {
+  char * key;
+  char * value;
+} rmw_qos_profile_payload_item_t;
+
+typedef struct rmw_qos_profile_payload_t {
+  rmw_qos_profile_payload_item_t * key_value_pairs;
+} rmw_qos_profile_payload_t;
+```
+
+It would be up to the implementation what to do with the passed key value pairs.
+
+Note:
+  The proposed API is limited, if we would want for example to define vendor specific QoS inline in the ROS QoS profile file.
```

Comment:
The limitation is subtle.

If the QoS policy can be represented by an integer or a float (e.g. history_depth is an integer), if we limit values to be a string. They won't be represented in the most natural way possible (specially, if the chosen format is YAML).
Support for a `variant` in C is painful, though possible.

Reply:
Makes sense. I guess it's first sentence that is confusing me. Can we simplify these two sentences to one? e.g.

> The proposed API is limited in that it treats all value as strings instead of taking advantage of natural representations supported by some markup languages.

Reply:
Done in https://github.com/ros2/design/pull/296/commits/c7ff192d83eea4e9c096389103a0bee4e27bb209.